pdl_schema.md

```diff
@@ -0,0 +1,1106 @@
+
+```pdl
+namespace com.example.models
+
+/**
+ * A common record to represent audit stamps.
+ */
+record AuditStamp {
+  /** Time when this record was created */
+  createdAt: Time
+
+  /** Time when this record was last updated */
+  updatedAt: Time
+}
+```
+A typeref allows additional meta-data to be associated with primitive and
+unnamed types. This meta-data can be used to provide documentation or support
+custom properties.
+
+### Typerefs can be used to name anonymous types.
+A typeref provides a way to refer to common unnamed types such as arrays, maps,
+and unions. Without typerefs, users may have to wrap these unnamed types with a
+record in order to address them. Alternatively, users may cut-and-paste common
+type declarations, resulting in unnecessary duplication and potentially causing
+inconsistencies if future changes are not propagated correctly to all copies.
+
+For example:
+```pdl
+namespace com.example.models
+
+typeref PhoneNumber = union[
+  /** A mobile phone number */
+  mobile: PhoneNumber,
+
+  /**
+   * A work phone number
+   */
+  work: PhoneNumber,
+
+  /** A home phone number */
+  home: PhoneNumber
+]
+
+```
+
+Typerefs can then be referred to by name from any other type:
+```pdl
+namespace com.example.models
+
+record Contacts {
+  primaryPhone: PhoneNumber
+  secondaryPhone: PhoneNumber
+}
+```
+
+### Typerefs can be used to specify custom types and coercers
+
+For example, Joda time has a convenient `DateTime` Java class. If we wish to use
+this class in Java to represent date-times, all we need to do is define a
+Pegasus custom type that binds to it:
+
+```pdl
+namespace com.example.models
+
+@java.class = "org.joda.time.DateTime"
+@java.coercerClass = "com.example.time.DateTimeCoercer"
+typeref DateTime = string
+```
+The coercer is responsible for converting the Pegasus “referenced” type, in this
+case "string" to the Joda `DateTime` class.
+
+See [Java Binding]
+(https://linkedin.github.io/rest.li/java_binding#custom-java-class-binding-for-primitive-types)
+for more details on defining and using custom types.
+
+## Fixed Type
+
+The Fixed type is used to define schemas with a fixed size in terms of number of
+bytes.
+
+For example:
+```pdl
+namespace com.example.models
+
+fixed MD5 16
+```
+The example above defines a type called `MD5`, which has a size of `16 bytes`.
+
+## Import
+Imports are optional statements which allow you to avoid writing fully-qualified
+names. They function similarly to imports in Java.
+
+For example, the following record can be expressed _without_ imports like this:
+
+```pdl
+namespace com.example.models
+
+record AuditStamp {
+  createdAt: com.example.models.time.Time
+  updatedAt: com.example.models.time.Time
+}
+```
+
+Alternatively, the record can be expressed using imports, minimizing the need
+for repetitive code:
+
+```pdl
+namespace com.example.models
+
+import com.example.models.time.Time
+
+record AuditStamp {
+  createdAt: Time
+  updatedAt: Time
+}
+```
+
+**Note:**
```

Comment:
Maybe be worthwhile to mention that import statements take precedence in name resolution.